tiny_dnn/util/math_functions.h

```diff
@@ -36,52 +37,56 @@ inline void vector_div(vec_t& x, float_t denom) {
 }
 
-/** 
- * calculate mean/variance across channels
- */
-inline void moments(const tensor_t& in, serial_size_t spatial_dim, serial_size_t channels, vec_t *mean, vec_t *variance) {
-    serial_size_t num_examples = static_cast<serial_size_t>(in.size());
-
-    assert(in[0].size() == spatial_dim * channels);
-
-    mean->resize(channels);
-    std::fill(mean->begin(), mean->end(), (float_t)0.0);
-
-    if (variance != nullptr) {
-        variance->resize(channels);
-        std::fill(variance->begin(), variance->end(), (float_t)0.0);
-    }
-
-    // calculate mean
-    for (serial_size_t i = 0; i < num_examples; i++) {
-        for (serial_size_t j = 0; j < channels; j++) {
-            float_t*       pmean = &mean->at(j);
-            const float_t* X = &in[i][j*spatial_dim];
-
-            for (serial_size_t k = 0; k < spatial_dim; k++) {
-                *pmean += *X++;
+namespace detail {
+    inline void moments_impl_calc_mean(serial_size_t num_examples, serial_size_t channels,
+                                       serial_size_t spatial_dim, const tensor_t& in, vec_t& mean) {
+        for (serial_size_t i = 0; i < num_examples; i++) {
+            for (serial_size_t j = 0; j < channels; j++) {
+                float_t& rmean = mean.at(j);
+                const auto it = in[i].begin() + (j * spatial_dim);
+                rmean = std::accumulate(it, it + spatial_dim, rmean);
             }
         }
     }
-
-    vector_div(*mean, (float_t)num_examples*spatial_dim);
-
-    // calculate variance
-    if (variance != nullptr) {
+    inline void moments_impl_calc_variance(serial_size_t num_examples, serial_size_t channels,
+                                           serial_size_t spatial_dim,
+                                           const tensor_t& in, vec_t& mean, vec_t& variance) {
         for (serial_size_t i = 0; i < num_examples; i++) {
             for (serial_size_t j = 0; j < channels; j++) {
-                float_t* pvar = &variance->at(j);
-                const float_t* X = &in[i][j*spatial_dim];
-                float_t        EX = (*mean)[j];
-
-                for (serial_size_t k = 0; k < spatial_dim; k++) {
-                    *pvar += pow(*X++ - EX, (float_t)2.0);
-                }
+                float_t& rvar = variance.at(j);
+                const auto it = in[i].begin() + (j * spatial_dim);
+                const float_t ex = mean[j];
+                rvar = std::accumulate(it, it + spatial_dim, rvar, [ex](float_t current, float_t x) {
+                    return current + pow(x - ex, (float_t)2.0);
+                });
             }
         }
-
-        vector_div(*variance, std::max(1.0f, num_examples*spatial_dim-1.0f));
+        vector_div(variance, std::max(1.0f, num_examples*spatial_dim-1.0f));
     }
 }
+/**
+ * calculate mean/variance across channels
+ */
+inline void moments(const tensor_t& in, serial_size_t spatial_dim, serial_size_t channels, vec_t& mean) {
```

Comment:
@beru This is discussed in https://github.com/tiny-dnn/tiny-dnn/pull/453 . IMHO 2-space + 80 char restriction is OK, though is not best for modern displays, 2/4-space with 120 char is the best.